Track status of the batteries

ela-kotulska-frequenz opened the pull request:
Add BatteryPoolStatus that tracks what batteries in pool are working or not.
Use BatteryPoolStatus in PowerDistributingActor temporary. 
In the future it will be used in BatteryPool.

ela-kotulska-frequenz committed:
Add more metrics to ComponentData

Add battery: relay state, component state, errors.
Add inverter: relay state, component state, errors.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Use BatteryPoolStatus in PowerDistribution

Replace old tool to track broken batteries with
new BatteryPoolStatus.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

tiyash-basu-frequenz commented on src/frequenz/sdk/microgrid/component/_component_errors.py:
```diff
@@ -0,0 +1,151 @@
+class BatteryErrorCode(Enum):
+    """Battery error codes."""
+
+    ERROR_CODE_UNSPECIFIED = battery_pb.ErrorCode.ERROR_CODE_UNSPECIFIED
```
This looks like a lot of duplication from protobuf here.
Will it be possible to just re-export the enum `battery_pb.ErrorCode` as `BatteryErrorCode`?

This comment extends also to other code in the files `_component_errors.py` and `_component_states.py`.

ela-kotulska-frequenz replied:
class name is BatteryErrorCode.
I would have to rename it to BatteryErrorCodePb or battery_error_code_pb. That is not very readable, too.

The modules could be separated into Battery, Inverter, Common (as in frequenz-api).
Now they are separated into component_erros, component_states etc.
But it is not discussed yet.

tiyash-basu-frequenz replied:
Not sure why we'd need to rename it to BatteryErrorCodePb or battery_error_code_pb. The enums being created here have the same variants as the protobuf ones, so we could just re-use the protobuf enums directly. (maybe I am missing something)

Couldn't we do something like the following?
```
from frequenz.api.microgrid.battery_pb2 import ErrorCodes as BatteryErrorCodes
```

Or is the actual intention here to get rid of the prefixes in the variants (`COMPONENT_STATE_`, `ERROR_CODE_`, etc.)?

shsms replied:
no,  we just don't want to expose protobuf generated code to users of the sdk.

tiyash-basu-frequenz replied:
> no, we just don't want to expose protobuf generated code to users of the sdk.

This decision is being enforced too strictly, and is causing us to duplicate enums.
I feel this is part of a broader discussion. I will suggest discussing this in a different channel, unblocking this PR.

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -0,0 +1,364 @@
+    _battery_invalid_state: Set[BatteryState] = {BatteryState.COMPONENT_STATE_ERROR}
+    _inverter_invalid_state: Set[InverterState] = {
+        InverterState.COMPONENT_STATE_UNSPECIFIED,
+        InverterState.COMPONENT_STATE_UNAVAILABLE,
```
Here are the valid states:
`BatteryRelayState`s: `CLOSED`
`BatteryState`s: `IDLE`, `CHARGING`, `DISCHARGING`
`InverterState`s: `STANDBY`, `IDLE`, `CHARGING`, `DISCHARGING`

You could either specify the complements of these as invalid states, or directly specify these as valid states. I think it's easier if you go with the latter, because the list of valid states is more stable than the list of invalid states.

ela-kotulska-frequenz replied:
done

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -0,0 +1,364 @@
+        self._run_method_called = False
+        # Time needed to get first component data.
+        # Constant for now. In future should be configurable in UI.
+        self._start_delay: float = 2.0
```
I'd suggest using a constant instead of `2.0`.  That way you can keep track of all the constants being used in this module in a single place.

ela-kotulska-frequenz replied:
done

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -0,0 +1,364 @@
+_logger = logging.getLogger(__name__)
+
+
+class BatteryStatus:
```
To me this looks like a collection of methods belonging to two different categories: ones that work on a single component, and ones that work on a collection of components.

So, I'd suggest splitting this into two classes for cleaner layers:
`BatteryPool`: has methods that work on a collection of battery IDs
`BatteryUtils`: has functions for  a single battery ID, like whether it's online or not, what is it's connected inverter, etc.

ela-kotulska-frequenz replied:
done

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -0,0 +1,364 @@
+        await asyncio.sleep(self._start_delay)
+        self._run_method_called = True
+
+    def _is_msg_correct(
```
I'd rename this to `_is_battery_ready`.

You could also alter the method signature to have the battery ID instead of battery and inverter data samples, and just get the status of the inverter connected to the battery.

ela-kotulska-frequenz replied:
done

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_status.py:
```diff
@@ -0,0 +1,364 @@
+            for bat_id in result.succeed_batteries:
+                self._blocked_batteries.unblock(bat_id)
+
+    def _is_msg_too_old(
```
Maybe `_is_data_outdated`?

ela-kotulska-frequenz replied:
done

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_status_utils.py:
```diff
@@ -0,0 +1,131 @@
+        return self.expiration_time <= now
+
+
+class BlockedComponents:
```
Maybe it makes sense to add this to component-specific classes, instead of being separate?

Then you could run functions like
```
if battery_0.is_blocked():
    bla()
```

You could also add a method `blocked_until()`, which returns the timestamp until when the component is blocked.

ela-kotulska-frequenz replied:
Done

tiyash-basu-frequenz reviewed (commented):
As suggested, reviewed from commit #5 onwards.

ela-kotulska-frequenz committed:
Add battery status module to track status of the battery

This should be used to check if data from the battery
should be used in computing and if battery should be used to
set power.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add BatteryPoolStatus module to store status for set of batteries

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

shsms commented on src/frequenz/sdk/microgrid/component/_component_data.py:
```diff
@@ -128,6 +134,11 @@ class BatteryData(ComponentData):
             charging is possible.
         temperature_max: the maximum temperature of all the blocks in a battery, in
             Celcius (°C).
+        relay_state: state of the battery relays. Each Relay is described in the
```
`Each *state* is ...`?

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/microgrid/component/_component_errors.py:
```diff
@@ -0,0 +1,151 @@
+
+    ERROR_LEVEL_UNSPECIFIED = common_pb.ErrorLevel.ERROR_LEVEL_UNSPECIFIED
+    ERROR_LEVEL_WARN = common_pb.ErrorLevel.ERROR_LEVEL_WARN
+    ERROR_LEVEL_CRITICAL = common_pb.ErrorLevel.ERROR_LEVEL_CRITICAL
```
we don't need to have the `ERROR_LEVEL_` prefix in our enums.  That would be consistent with how it is done in other such enums as well.

ela-kotulska-frequenz replied:
Done

shsms replied:
there are a few in `_component_states.py` as well.

ela-kotulska-frequenz replied:
Thanks, done! :)

shsms commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,121 @@
+    ) -> None:
+        """Create object instance.
+
+        Note:
+            The object instance is fully created after running `run` method.
```
Another option is to have a `BatteryPool.new()` async class method that creates and returns a BatteryPool instance.  And in the class documentation say "use `pool = await BatteryPool.new(...)` to create".

``` python
from __future__ import annotations

class BatteryPool:
    """Use new() to create."""
    _CreatedThroughNew = object()

    def __init__(..., created_by: Any = None):
        """Not for direct use.  Use new() to create."""
        if created_by is not _CreatedThroughNew:
             raise RuntimeError(...)
        ... rest of init

    async def new(...) -> BatteryPool:
        await etc
        return BatteryPool(...., cls._CreatedThroughNew)

```

ela-kotulska-frequenz replied:
Yes but then it will be impossible to create BatteryPool in any constructor.
So user would have to either:
1) In constructor define battery pool as: `battery_pool: Optional[BatteryPool] = None` -> Create battery pool in other(async) method -> before every use check None condition. (which would end up (in most cases) with ignoring pylint/mypy warnings)
2) Define BatteryPool outside constructor and ingore mypy/pylint warnings that object is defined outside `__init__`

So from usage perspective it is better to define it in this way :(
It is true , that user has to remember about `run` method. But at least it is easy to use.

shsms replied:
True,  okay. :-)

Then what do you think about the name `async_init` (or `init_batteries`, etc)?  `run` sounds like it would start a long running task,  but that's not the case.

ela-kotulska-frequenz replied:
I go with `init` because:
* `async_init` would look like: `async def async_init(self)` . 
* `init` is very close to `__init__` to so it makes sense it is part of the init step.
*   `init_batteries` - it is good, but can be misunderstood. User might wonder why he has to init batteries etc.
* BatteryPool is private class, now. We can rename `init` to anything else when we publish. And until we publish it,  it will be no harm to user :)

shsms replied:
The only issue is that now when I say "the init method", it is ambiguous which one I mean.

We'll probably end up calling it the "async init" anyway.

> `async_init` would look like: `async def async_init(self)` .

Is that bad?

ela-kotulska-frequenz replied:
async before function definition tells that this method is async.

ela-kotulska-frequenz replied:
Ok I will rename it to async_init.

`async def async_init(self) ` - no, it is just lots of `async`

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz replied:
Damn GitHub, it lost my comment when I hit unresolve conversation :sob: 

I tried to say that I don't fully agree with not being able to use this class as other class attributes in constructors. There is another option, which is what happens in general with asyc. Async is viral, and if you want to use it, you need to make your stuff async and IMHO this should happen with constructors too, otherwise we are just making users need to write more error prone code.

So IMHO users that want to have a `BatteryPool` in their class should make their class *async-constructible* too, otherwise they will also need to split sync construction from async construction. We can add a utility class `AsyncConstructible` to allow users to also implement their own async constructible classes more easily, for example:

```py
class NotSyncConstructible(AssertError):
    pass


class AsyncConstructible(ABC):
    def __init__(self) -> None:
        raise NotSyncConstructible()

    @abstractmethod
    @classmethod
    async def async_new(cls) -> Self:  # or AsyncConstructible if it still doesn't work
        return cls.__new__(cls)


class BatteryPool(AsyncConstructible):
    def __init__(self) -> None:
        self.attr1: str
        """Documentation for attr1."""
        self.attr2: int
        """Documentation for attr2."""
        
        super().__init__()

    @classmethod
    def async_new(cls, ...) -> Self:  # or BatteryPool if it still doesn't work
        self = await super().async_new()
        self.attr1 = "lala"
        self.attr2 = 2
        return self
```

We just need to check if we can somehow document instance attributes.

All that said, I'm also fine to leave this PR as is and create a new issue for this, as it is not strictly related to the PR and it might need more discussion.

PS: I know this also add some error proneness, as users need to make sure to define their classes correctly, but at least you need to define a class (and get it right) only once, while if the error proneness is in the usage, you have to do it several times, having more chances to mess it up.

ela-kotulska-frequenz replied:
```
class BatteryPool(AsyncConstructible):
    @classmethod
    def async_new(cls, ...) -> Self:  # or BatteryPool if it still doesn't work
        self = await super().__async_new()
        # Assign attributes, do more async stuff
        return self
```
About line     # Assign attributes, do more async stuff
If attribute was not assigned in method with self argument, then pylint doesn't recognize this attribute.  So there is lots of pylint errors that we have to quiet to make it work.

We should skip this topic in this PR and continue in different issue. 
https://github.com/frequenz-floss/frequenz-sdk-python/issues/135

shsms commented:
I guess the PR title needs to be changed as well?

shsms commented on src/frequenz/sdk/microgrid/component/_component_states.py:
```diff
@@ -39,3 +41,83 @@ def from_pb(
+class BatteryRelayState(Enum):
+    """Battery relay states."""
+
+    RELAY_STATE_UNSPECIFIED = battery_pb.RelayState.RELAY_STATE_UNSPECIFIED
+    RELAY_STATE_OPENED = battery_pb.RelayState.RELAY_STATE_OPENED
+    RELAY_STATE_PRECHARGING = battery_pb.RelayState.RELAY_STATE_PRECHARGING
+    RELAY_STATE_CLOSED = battery_pb.RelayState.RELAY_STATE_CLOSED
+    RELAY_STATE_ERROR = battery_pb.RelayState.RELAY_STATE_ERROR
+    RELAY_STATE_LOCKED = battery_pb.RelayState.RELAY_STATE_LOCKED
```
here too, I guess the prefix can go away.

ela-kotulska-frequenz replied:
No enums anymore

shsms reviewed (commented):
Just looked at the first commit again.  Will finish review tomorrow.

tiyash-basu-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,334 @@
+
+        return is_ready
+
+    def is_blocked(self) -> bool:
```
From an UX perspective, it would seem that `!battery.is_ready()` would achieve the same purpose. But it would not, therefore introducing a bit of inconsistency.

Looks like the timestamp check is not a part of the `is_ready` method. If you add it there, then the `is_blocked` method can be removed.

ela-kotulska-frequenz replied:
But then I can't get batteries that are blocked and ready.

ela-kotulska-frequenz replied:
I can't just move it, because then I won't be able to get batteries with correct message but are blocked.

I renamed `is_ready` to `get_state` that returns ENUM with information if battery is working, blocked or not working.

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,122 @@
+        if len(working) > 0:
+            return working
+
+        _logger.warning("All batteries in %s are not working. Use blocked batteries.")
```
The part `Use blocked batteries.` sounds like an instruction, and can be confusing. What is the intention here?

shsms replied:
Maybe something like "There are no working batteries in %s.  Falling back to using blocked batteries."?

ela-kotulska-frequenz replied:
Done

tiyash-basu-frequenz commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,122 @@
+            for bat_id in result.failed_batteries:
+                duration = self._batteries[bat_id].block()
+                if duration > 0:
+                    _logger.info(
```
I'd suggest setting the log level to `warn` for this message.

ela-kotulska-frequenz replied:
Done

tiyash-basu-frequenz reviewed (commented):
Thanks for addressing the last set of comments. I have a few more.

matthias-wende-frequenz commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+
+
+class BatteryPool:
+    """Holds bool of batteries and returns data from them."""
```
pool ;)

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+        await asyncio.sleep(_START_DELAY_SEC)
+        self._init_method_called = True
+
+    def get_state(self) -> WorkingState:
+        """Return if battery message indicates that battery can be used.
+
+        Raises:
+            RuntimeError: if method `init` was not called or not awaited.
+
+        Returns:
+            True if battery can be used, False otherwise.
+        """
```
The documentation seems to suggest that this returns a bool?

ela-kotulska-frequenz replied:
Done

shsms commented on tests/actor/test_battery_pool.py:
```diff
@@ -0,0 +1,929 @@
+import frequenz.api.microgrid.microgrid_pb2 as microgrid_pb
+import pytz
+import time_machine
+from frequenz.api.microgrid.battery_pb2 import Battery as PbBatteryData
```
I think these should be aliased just as `PbBattery` for consistency,  because there's `from ...battery_pb2 import Data as PbBatteryData` somewhere else.

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+        self._run_method_called: bool = False
+
+    async def init(self) -> None:
+        """Run battery pool."""
```
Just `init`, not `run`, I guess?

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        if not self._run_method_called:
+            raise RuntimeError(
+                "`run` method not called or not awaited. Run it before first use"
```
`_init_called`?  and `init method not called ...`.  In the docstring also.

leandro-lucarella-frequenz replied:
Making instantiation and async init one step will also avoid having to do all these checks (or make them just an `assert` as users would have to be doing something really weird to get into a inconsistent state).

ela-kotulska-frequenz replied:
@leandro-lucarella-frequenz please read answer below :) 
> Making instantiation and async init one step will also avoid having to do all these checks 

https://github.com/frequenz-floss/frequenz-sdk-python/pull/117#discussion_r1050787860

ela-kotulska-frequenz replied:
> _init_called? and init method not called .... In the docstring also.

Done

shsms commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+        """Create partially initialized object instance.
+
+        Note:
+            Please call `run` method to fully initialize BatteryPool. Otherwise
```
`init`

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        if not self._run_method_called:
+            raise RuntimeError(
+                "`run` method not called or not awaited. Run it before first use"
```
init

ela-kotulska-frequenz replied:
done

shsms reviewed (commented):
I think these classes should be called something else.  We have already reserved the name `BatteryPool` for the other thing that runs formulas on resampled data and sits in the `timeseries` module.

The class in this PR is really a helper class for the PowerDistributor, that operates on direct component data,  and is tightly coupled with the PowerDistributor.

These classes are really just about `Status` like the original name suggested,  and we might want to retain that word in the names.   Maybe something like "BatteryStatus" and "BatteryPoolStatus".

leandro-lucarella-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+"""Class to store features for battery."""
```
The `_data_ingestion` package is supposed to die, what about putting this in `microgrid` instead? Also, what about making it private (`_battery_utils.py`) if it's not intended to be used by SDK users directly (or is it?). While we are at it, would it make sense to have a `battery` package (if we are going to add more modules related to battery handling)? Like `sdk.microgrid.battery.util`?

ela-kotulska-frequenz replied:
Changed to `_status.py` and moved to `sdk.microgrid._battery._status`
New name for the class `StatusTracker`

leandro-lucarella-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+    receiver: Peekable[T]
+
+
+class BatteryUtils:
```
I'm not crazy about this name. Isn't this mainly keeping track of the status of the battery? Why not `BatteryStatus`? Alternatives: `BatteryController`, `BatteryBlocker`, *util* is way too ambiguous and general IMHO.

shsms replied:
For the other class also, it is also not really the data-pipeline BatteryPool that we have been talking about,  but something specific to the the PowerDistributor: https://github.com/frequenz-floss/frequenz-sdk-python/pull/117#pullrequestreview-1227363605

leandro-lucarella-frequenz replied:
If it is stuff internal to the power distributor, then I think it should live in `sdk.power` (or `actor.power_distribution` if this really is only intended to be used by the actor and nobody else) as internal modules `_battery_status.py` and `_battery_pool.py`. For me it is still OK to use the name if it makes sense, but it should be definitely very clear that it is some internal implementation detail of the power distributor then.

shsms replied:
No reason to reuse names and introduce ambiguities when we don't have to.

ela-kotulska-frequenz replied:
BatteryPool is now used in PowerDistributor. But in the future it won't.

PowerDistirbutor will not track what batteries are working in future. It will just distribute power over given batteries.
User will use BatteryPool.set_power(...). And BatteryPool will use PowerDistributor.

ela-kotulska-frequenz replied:
BatteryPool is used in PowerDistributor temporary. It will be extended to use the formulas and exposed to user.

leandro-lucarella-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+            return False
+        return self._blocked_until > datetime.now(tz=pytz.UTC)
+
+    def unblock(self):
```
`-> None`, weird if it's not failing the linting...

ela-kotulska-frequenz replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+
+        return self._last_blocking_duration_sec
+
+    def blocked_until(self) -> Optional[datetime]:
```
`@property`?

ela-kotulska-frequenz replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+
+        Returns:
+            Timestamp when the battery will be unblocked. Return None if battery is
+            not blocked.
```
Needs extra indentation to continue the line:
```py
        Returns:
            Timestamp when the battery will be unblocked. Return None if battery is
                not blocked.
```

ela-kotulska-frequenz replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/_data_ingestion/battery_utils.py:
```diff
@@ -0,0 +1,361 @@
+
+        return self._last_state
+
+    def _is_blocked(self) -> bool:
```
Wouldn't it make sense to make this a public property? I know there is `blocked_util` too, but if you just want to know if it's blocked or not it feels weirder to do:
```py
if b.blocked_until is not None:
    print("is blocked")
```

ela-kotulska-frequenz replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+_logger = logging.getLogger(__name__)
+
+
+class BatteryPool:
```
(Small typo in the commit message: ***Int** the future it will be an actor that streams*)

I would put this class in `sdk.microgrid.battery` instead, as it is not an actor. IMHO actors should only be a thin wrapper over other reusable modules in the SDK.

Also, we need to start asking the question, is this class intended to be used by SDK users or it is internal, because if it internal it should be in a private module (starting with `_`).

ela-kotulska-frequenz replied:
Yes, but now it needs PowerDistirbutingResult. And PowerDistributing needs BatteryPool. And that is circular dependency  :( 
I the future it will be moved.
I will make it clear in commit message.

ela-kotulska-frequenz replied:
> Also, we need to start asking the question, is this class intended to be used by SDK users or it is internal, because if it internal it should be in a private module (starting with _).

It is intended to be used by users in the future

leandro-lucarella-frequenz replied:
OK, as long as you are having what should be public interface and what should be internal and hidden, all good if we need to go through an intermediary step where things are not where they should in a final version :+1:

leandro-lucarella-frequenz replied:
(OTOH as long as stuff is temporary I would still make the module hidden to make it clear for users that if they use that module it can disappear without notice at any time)

ela-kotulska-frequenz replied:
OK I made it private

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/battery_pool.py:
```diff
@@ -0,0 +1,125 @@
+        }
+        self._run_method_called: bool = False
+
+    async def init(self) -> None:
```
I've seen this pattern twice now, and I wonder if users will **always** need to do:

```py
c = AsyncClass(...)
await c.init()
```

Then we should probably add an alternative async constructor instead:

```py

@classmethod
async def new(cls, ...) -> Self:  # Maybe this didn't work yet, even with typing_extensions, but we can use the concrete class for now
    o = cls(...)
    await o._init()
    return o
```

We could take it even further and forbid using the constructor:

```py
def __init__(self):
    # For documentation purposes maybe we can declare the instance attribute types
    raise NotImplementedError(f"Use {self.__class__.__name__}.new() instead")

@classmethod
async def new(cls, ...) -> Self:  # Maybe this didn't work yet, even with typing_extensions, but we can use the concrete class for now
    o = cls.__new__(cls)
    # initialize attributes
    await o._init()
    return o
```

Just an idea, but it would make it less error prone, not giving users the possibility to forget to call `init()`

ela-kotulska-frequenz replied:
Yes we discussed it: and it is better to do it this way.
https://github.com/frequenz-floss/frequenz-sdk-python/pull/117#discussion_r1050787860

leandro-lucarella-frequenz replied:
Ah, sorry I missed the conversation there. I will continue the discussion there.

leandro-lucarella-frequenz replied:
Ah, sorry I missed the conversation there. I will continue the discussion there.

leandro-lucarella-frequenz replied:
Ah, sorry I missed the conversation there

ela-kotulska-frequenz commented:
Renamed BatteryPool to BatteryPoolStatus as discussed on the meeting.

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/_component_data.py:
```diff
@@ -128,6 +130,9 @@ class BatteryData(ComponentData):
             Celcius (°C).
+        _relay_state: state of the battery relay.
+        _component_state: state of the battery.
+        _errors: list of errors in protobuf struct.
```
It is OK for this PR, as there were already attributes documented like this, but in the future we need to move from documenting in `Attributes:` to documenting as docstrings where the attributes are defined, as the documentation tool we use renders the attributes much better if they are documented there.

ela-kotulska-frequenz replied:
Done (last commit).

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/_component_data.py:
```diff
@@ -137,6 +142,9 @@ class BatteryData(ComponentData):
     power_lower_bound: float
     power_upper_bound: float
     temperature_max: float
+    _relay_state: battery_pb.RelayState.ValueType
+    _component_state: battery_pb.ComponentState.ValueType
+    _errors: List[battery_pb.Error]
```
So here, like:
```py
    _relay_state: battery_pb.RelayState.ValueType
    """State of the battery relay."""
    _component_state: battery_pb.ComponentState.ValueType
    """State of the battery."""
    _errors: List[battery_pb.Error]
    """List of errors in protobuf struct."""
```

With dataclasses it will say they are class attributes, but it is also marked as a dataclass so it is still easy for users to know they are really instance attributes.

ela-kotulska-frequenz replied:
Done (last commit).

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -0,0 +1,373 @@
+        """
+        if self._blocked_until is None:
+            return False
+        return self._blocked_until > datetime.now(tz=pytz.UTC)
```
Nitpick: I would avoid using `pytz` only for `UTC`, as it is available in the stdlib as `datetime.timezone.utc`.

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_pool_status.py:
```diff
@@ -0,0 +1,126 @@
+
+    async def async_init(self) -> None:
+        """Init battery pool."""
+        tasks = [
+            asyncio.create_task(battery.async_init())
+            for battery in self._batteries.values()
+        ]
+        await asyncio.wait(tasks, return_when=ALL_COMPLETED)
```
Minor: I guess it could be jusut `asynio.gather()`, right?

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on tests/actor/test_power_distributing.py:
```diff
@@ -232,14 +229,23 @@ async def test_power_distributor_one_user(
             )
 
+            # Mock that all requested batteries are working.
+            distributor._battery_pool = MagicMock()
```
Something that we are not using very often but we probably should is the `spec` argument for mocks, it will validate that the mock is only used via attributes or calls that follow some object or class. For example: `distributor._battery_pool = MagicMock(spec=BatteryPoolStatus)` should create a mock object that already has a `async_init()` attribute and it is an `AsyncMock` because it knows the signature of the method, and will also validate that the arguments of the method follow the specs.

ela-kotulska-frequenz replied:
I didn't know that. Thanks, done

leandro-lucarella-frequenz reviewed (commented):
Some final minor comments. If they are addressed, great but otherwise from my side this can go as it is too. I will leave the approval pending because there were other reviewers that had comments.

ela-kotulska-frequenz committed:
Fix documentation in component_data module

Docstrings attributes should not be documented in `Attributes`.
We should document attribute below its definition, because
the documentation tool we use renders the attributes much better
if they are documented that way.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

sahas-subramanian-frequenz reviewed (commented):
LGTM,  I can approve once it is rebased.
